jobrunner/executors/local.py

```diff
@@ -279,12 +283,15 @@ def get_status(self, job_definition, timeout=15):
         metrics = record_stats.read_job_metrics(job_definition.id)
 
         if container is None:  # container doesn't exist
-            if job_definition.cancelled:
+            # cancelled=True indicates that we are in the process of cancelling this
+            # job. If we're not, the job may have been previously cancelled; look up
+            # its cancelled status in job metadata, if it exists
+            if cancelled or (self.get_metadata(job_definition) or {}).get("cancelled"):
```

Comment:
Do you think that we could avoid having to pass cancelled to get_status if

a) in finalize_job, we add a cancelled key to job_metadata, and set to to value of the cancelled variable we passed to finalize_job?
b) in get_status, we read the metadata earlier, and use metadata["cancelled"] to detect cancellation?

This also permanently persists the metadata that this job was cancelled

wydt?

Reply:
I don't know that we can; we call get_status() as one of the first things in `handle_canceljob_task`; the status we get back there depends on whether we're cancelled or not. When we're at the start of `handle_canceljob_task`, we haven't written any metadata yet, because we haven't actually cancelled it yet, so we haven't called `finalize`

Reply:
And it's already doing (a)

Reply:
Hmm, yeah, I see the problem.

So, I think it's possible that we do not need get_status correctly differentiate between PREPARED and UNKNOWN in the face of being cancelled anymore, as we're calling it from an entirely separate task handler. The only reason I think it currently exists is to set a different message, which is no longer the agents responsibility.

So, I think we could possibly remove all cancellation special casing in get_status. That will still lead to UNKNOWN or PREPARED being returned, which are handled appropriately already? There is no message to set.

In fact, it might be worth getting rid of JobStatus.message altogether?